cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/LinkAnalysis.scala

```diff
@@ -0,0 +1,121 @@
+   */
+  protected def start: T
+
+  /**
+   * Calculate the error between two arrays using either the T1 error or the T2 error
+   * @param t1 Flag `true` to calculate the T1 error (the sum of absolute differences between two arrays).  Flag
+   *           `false` to calculate the T2 error (the sum of the squared differences between two arrays).
+   */
+  protected def error(a: Array[Double], b: Array[Double], t1: Boolean = true): Double = {
+    val rawError = (a zip b).map { case (v1,v2) => if(t1) Math.abs(v1 - v2) else Math.pow(v1 - v2, 2.0) }.sum
+    if (t1) rawError else Math.sqrt(rawError)
+  }
+}
+
+/**
+ * The base class for all parameters fed to our iterative algorithms.
+ * @param maxIterations The maximum number of times that the link analysis algorithm will run before termination
+ * @param tolerance The maximum error allowed.
+ */
+abstract class Params(val maxIterations: Option[Int], val tolerance: Double)
+
+/**
+ * The base class for all iterations through our iterative algorithms.  These classes will simply hold all of the
+ * information needed to assess the number of iterations, the error, and the current set of values.
+ * @param error  The T1 error for the current iteration vs the previous iteration.
+ */
+abstract class Iteration(val error: Double, val iteration: Int)
+
+/**
+ * `AbstractLinkAnalysis` is the base class that extends `LinkAnalysis`.  All algorithms should inherit from
+ * this class.  Let it be noted that we exploit scala streams to accomplish the recursion necessary for
+ * our algorithms.  We will not be storing the results every step of the way--instead we find the iteration
+ * that first satisfies the condition of exceeding maximum iterations or falls below the error threshold.
+ * @param params The set of all parameters passed into our algorithm
+ * @param modelName This must be provided by all subclasses for logging purposes.  This should not be
+ *                  left up to the user to define.  This must be provided by the implementer.
+ * @param initial  The starting iteration for our algorithms.  This can be provided by the user or can be
+ *                 given the default value provided by the `start` method.
+ * @tparam T `LinkAnalysis` must be generically typed by `Iteration` or one of its subclasses.  An `Iteration`
+ *           holds all of the pertinent information for a given algorithm.
+ */
+abstract class AbstractLinkAnalysis[T <: Iteration](graph: DirectedGraph[Node],
+    params: Params, modelName: String, initial: Option[T] = None) extends LinkAnalysis[T] {
+
+  protected val log = Logger.get(modelName)
+
+  val maxIterations = params.maxIterations
+  val tolerance     = params.tolerance
+
+  /**
+   * This is the final `Iteration` object from our stream.  The user can feed this iteration back to `iterate`
+   * or `run` if they wish to converge further.
+   */
+  val result = if ((maxIterations.isDefined && maxIterations.get > 0) || !maxIterations.isDefined)
```

Comment:
not sure what the benefit of evaluating result, error, iteration is. let the user call run() and get the final error and iterations used?